File: api/chat/retriever.py
from dataclasses import dataclass
import math
import re
from collections import Counter
from typing import List
import joblib

TOKEN_RE = re.compile(r"[a-z0-9_@./:+-]+")


def _tokenize(text: str):
    return TOKEN_RE.findall(text.lower())


def _dot(a, b):
    if len(a) > len(b):
        a, b = b, a
    return sum(v * b.get(k, 0.0) for k, v in a.items())


def _norm(a):
    return math.sqrt(sum(v * v for v in a.values()))


@dataclass
class RetrievalHit:
    id: str
    question: str
    answer: str
    source: str
    score: float


class Retriever:
    def __init__(self, index_path: str):
        payload = joblib.load(index_path)
        self.idf = payload["idf"]
        self.doc_vectors = payload["doc_vectors"]
        self.docs = payload["docs"]
        self.doc_norms = [_norm(v) for v in self.doc_vectors]

    def search(self, query: str, top_k: int = 4):
        q_tokens = _tokenize(query)
        q_tf = Counter(q_tokens)
        q_len = max(1, len(q_tokens))
        q_vec = {
            tok: (count / q_len) * self.idf.get(tok, 0.0)
            for tok, count in q_tf.items()
            if tok in self.idf
        }
        q_norm = _norm(q_vec)

        sims = []
        for d_vec, d_norm in zip(self.doc_vectors, self.doc_norms):
            if q_norm == 0.0 or d_norm == 0.0:
                sims.append(0.0)
            else:
                sims.append(_dot(q_vec, d_vec) / (q_norm * d_norm))

        top_idx = sorted(range(len(sims)), key=lambda i: sims[i], reverse=True)[:top_k]
        hits = []
        for idx in top_idx:
            doc = self.docs[idx]
            hits.append(
                RetrievalHit(
                    id=doc["id"],
                    question=doc["question"],
                    answer=doc["answer"],
                    source=doc["source"],
                    score=float(sims[idx]),
                )
            )
        return hits

File: src/lib/retriever.ts
export interface RetrievalHit {
  id: string;
  question: string;
  answer: string;
  source: string;
  score: number;
}

interface Doc {
  id: string;
  question: string;
  answer: string;
  source: string;
}

interface IndexPayload {
  idf: Record<string, number>;
  doc_vectors: Record<string, number>[];
  docs: Doc[];
}

const TOKEN_RE = /[a-z0-9_@./:+-]+/g;

function tokenize(text: string): string[] {
  return text.toLowerCase().match(TOKEN_RE) || [];
}

function dot(a: Record<string, number>, b: Record<string, number>): number {
  const keys = Object.keys(a).length > Object.keys(b).length ? Object.keys(b) : Object.keys(a);
  return keys.reduce((sum, k) => sum + (a[k] || 0) * (b[k] || 0), 0);
}

function norm(vec: Record<string, number>): number {
  return Math.sqrt(Object.values(vec).reduce((sum, v) => sum + v * v, 0));
}

class Retriever {
  private idf: Record<string, number>;
  private docVectors: Record<string, number>[];
  private docs: Doc[];
  private docNorms: number[];

  constructor(payload: IndexPayload) {
    this.idf = payload.idf;
    this.docVectors = payload.doc_vectors;
    this.docs = payload.docs;
    this.docNorms = this.docVectors.map(v => norm(v));
  }

  search(query: string, topK: number = 4): RetrievalHit[] {
    const qTokens = tokenize(query);
    const qTf: Record<string, number> = {};
    for (const tok of qTokens) {
      qTf[tok] = (qTf[tok] || 0) + 1;
    }
    const qLen = Math.max(1, qTokens.length);
    const qVec: Record<string, number> = {};
    for (const [tok, count] of Object.entries(qTf)) {
      if (tok in this.idf) {
        qVec[tok] = (count / qLen) * this.idf[tok];
      }
    }
    const qNormVal = norm(qVec);

    const sims: number[] = [];
    for (let i = 0; i < this.docVectors.length; i++) {
      const dVec = this.docVectors[i];
      const dNorm = this.docNorms[i];
      if (qNormVal === 0 || dNorm === 0) {
        sims.push(0);
      } else {
        sims.push(dot(qVec, dVec) / (qNormVal * dNorm));
      }
    }

    const topIdx = sims
      .map((score, idx) => ({ score, idx }))
      .sort((a, b) => b.score - a.score)
      .slice(0, topK)
      .map(x => x.idx);

    return topIdx.map(idx => ({
      id: this.docs[idx].id,
      question: this.docs[idx].question,
      answer: this.docs[idx].answer,
      source: this.docs[idx].source,
      score: sims[idx],
    }));
  }
}

let _retriever: Retriever | null = null;

export async function getRetriever(): Promise<Retriever> {
  if (!_retriever) {
    const res = await fetch('/index.json');
    const payload = await res.json() as IndexPayload;
    _retriever = new Retriever(payload);
  }
  return _retriever;
}

export async function searchFAQ(query: string, topK: number = 4): Promise<RetrievalHit[]> {
  const retriever = await getRetriever();
  return retriever.search(query, topK);
}

export function extractName(message: string): string | null {
  const patterns = [
    /my name is (\w+)/i,
    /i am (\w+)/i,
    /i'm (\w+)/i,
    /this is (\w+)/i,
    /call me (\w+)/i,
  ];
  for (const pattern of patterns) {
    const match = message.match(pattern);
    if (match) return match[1];
  }
  return null;
}
